### PR TITLE
Issue1-import-testapp1-code

### LINKED ISSUE #1: Import the testapp1 code
- Import the code for testing membrane-backend
- Ensure multiple instances can be spin up with different env vars
- Ensure that it's possible to enable / disable membrane

### PR DESCRIPTION
**Pull request by k-allagbe**
Closes #1
issue1-import-testapp1-code

Importation done. Refactored to remove code smells. Unit tests pass. Manual tests (auth scenario) done.

### CODE REVIEW

File: .env.template [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# Mandatory
+FASTAPI_EXAMPLE_SECRET_KEY=
```
Comment (rngadam) [inline .env.template]: FLASK_EXAMPLE_

File: .vscode/settings.json [hunk 2]
```diff
@@ -0,0 +1,13 @@
+{
+    "python.linting.enabled": true,
+    "python.testing.pytestArgs": [
+        "tests"
+    ],
+    "python.testing.unittestEnabled": false,
+    "python.testing.pytestEnabled": true,
```
Comment (rngadam) [inline .vscode/settings.json]: but we use unittest, why is this referencing pytest?

File: .vscode/settings.json [hunk 3]
```diff
@@ -0,0 +1,13 @@
+{
```
Comment (rngadam) [inline .vscode/settings.json:1]: if we decide to also add .vscode settings, then we should document it in dev-rel-docs

File: .vscode/settings.json [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+    "python.testing.pytestArgs": [
+        "tests"
+    ],
+    "python.testing.unittestEnabled": false,
+    "python.testing.pytestEnabled": true,
+    "[python]": {
+        "editor.defaultFormatter": "ms-python.black-formatter",
+        "editor.formatOnSave": true
+    },
+    "python.formatting.provider": "none"
```
Comment (rngadam) [inline .vscode/settings.json]: if we are going to do this, also do trimOnSave

File: app.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    if jwt_token:
+        decoded_token = decode_jwt_token(jwt_token, SERVER_PUBLIC_KEY)
+        if decoded_token:
+            session.permanent = True
+            session["authenticated"] = True
+            session["email"] = decoded_token["sub"]
+            return redirect(url_for("hello_world"))
+
+    # Fallback: Redirect to authentication
+    data = {"app_id": APP_ID, "redirect_url": url_for("hello_world", _external=True)}
```
Comment (rngadam) [inline app.py]: better name than hello_world (example_endpoint)

File: app.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    "FASTAPI_EXAMPLE_SERVER_PUBLIC_KEY_FILE", "keys/server_public_key.pem"
+)
+CLIENT_PRIVATE_KEY_FILE = os.getenv(
+    "FASTAPI_EXAMPLE_CLIENT_PRIVATE_KEY_FILE", "keys/client_private_key.pem"
+)
+SESSION_LIFETIME = int(os.getenv("FASTAPI_EXAMPLE_SESSION_LIFETIME", 300))
+TOKEN_EXPIRES_IN_SECONDS = int(
+    os.getenv("FASTAPI_EXAMPLE_TOKEN_EXPIRES_IN_SECONDS", 300)
+)
+MEMBRANE_SERVER = os.getenv("FASTAPI_EXAMPLE_MEMBRANE_SERVER", "http://127.0.0.1:5000")
```
Comment (rngadam) [inline app.py]: supposed to be optional but doesn't look like your code works without it?

File: app.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+
+@app.route("/health", methods=["GET"])
+def health():
+    return "ok", 200
+
+
+@app.route("/")
+def hello_world():
+    if session.get("authenticated"):
```
Comment (rngadam) [inline app.py]: see how they do it for flask-login:

https://www.digitalocean.com/community/tutorials/how-to-add-authentication-to-your-app-with-flask-login

...and do we want to extend flask-login to support membrane?

File: init-client.sh [hunk 8]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
```
Comment (rngadam) [inline init-client.sh]: init-client should be a script provided by membrane-backend (eventually membrane-client)

File: init-client.sh [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+
+set -e  # Exit on error
+set -x  # Debug mode
+
+# Create keys directory if it doesn't exist
+mkdir -p keys
+
+# Generate test keys
+openssl genpkey -algorithm RSA -out "tests/test_keys/private_key.pem" >> init.log 2>&1
```
Comment (rngadam) [inline init-client.sh]: if these are for testing, they should be generated in the tests

File: init-client.sh [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+openssl rsa -pubout -in "tests/test_keys/private_key.pem" -out "tests/test_keys/public_key.pem" >> init.log 2>&1
```
Comment (rngadam) [inline init-client.sh]: could we generate these keys to be in memory only (set in the environment variable?)

File: tests/test_jwt_functions.py [hunk 11]
```diff
@@ -0,0 +1,47 @@
+import unittest
+from datetime import datetime, timedelta
+
+import jwt
+
+from app import JWTDecodingError, decode_jwt_token, generate_jwt_token
```
Comment (rngadam) [inline tests/test_jwt_functions.py]: this test is a test that should only exists in membrane-backend... which we should depend on here.

File: tests/test_routes.py [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+    def test_health_route(self):
+        response = self.client.get("/health")
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response.data, b"ok")
+
+    def test_hello_world_authenticated(self):
+        # TODO
+        pass
+
+    def test_hello_world_no_token(self):
```
Comment (rngadam) [inline tests/test_routes.py]: also test when MEMBRANE_SERVER is not defined (should still work, just no authentication)

**Review by rngadam — CHANGES_REQUESTED** [review]

File: .env.template (hunk 1 above)
Comment (k-allagbe) [reply .env.template]: Done

File: .vscode/settings.json (hunk 2 above)
Comment (k-allagbe) [reply .vscode/settings.json]: Done

File: .vscode/settings.json (hunk 3 above)
Comment (k-allagbe) [reply .vscode/settings.json:1]: I created an issue for this: https://github.com/ai-cfia/dev-rel-docs/issues/40

File: .vscode/settings.json (hunk 4 above)
Comment (k-allagbe) [reply .vscode/settings.json]: The closest I found was trimTrailingWhitespace

File: app.py (hunk 5 above)
Comment (k-allagbe) [reply app.py]: Done

File: init-client.sh (hunk 8 above)
Comment (k-allagbe) [reply init-client.sh]: Removed altogether

File: init-client.sh (hunk 9 above)
Comment (k-allagbe) [reply init-client.sh]: Removed

File: init-client.sh (hunk 10 above)
Comment (k-allagbe) [reply init-client.sh]: Removed

File: tests/test_jwt_functions.py (hunk 11 above)
Comment (k-allagbe) [reply tests/test_jwt_functions.py]: Removed

File: app.py [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+    blueprint,
+    configure,
+    membrane_current_user,
+    membrane_login_required,
+)
+
+load_dotenv()
+
+# Constants
+ACTIVATE_LOGIN = os.getenv("FLASK_EXAMPLE_ACTIVATE_LOGIN", "").lower() == "true"
```
Comment (rngadam) [inline app.py]: Why not just look at whether MEMBRANE_SERVER is defined or not?

File: tests/test_routes.py [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+
+        response = self.client.get("/")
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(
+            response.data.decode("utf-8"), "Hello, test_email@example.com!"
+        )
+
+    @patch("app.membrane_login_required")
+    def test_example_endpoint_without_login(self, mock_login_required):
+        mock_login_required.side_effect = lambda func: func
```
Comment (rngadam) [inline tests/test_routes.py]: I'm confused as to what seems to be the purpose of the test (without_login) and how you've configured the annotations (current_user vs login_required). I would have expected login_required annotation to not be there?

File: app.py [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+SECRET_KEY = os.getenv("FLASK_EXAMPLE_SECRET_KEY")
+SESSION_LIFETIME = int(os.getenv("FLASK_EXAMPLE_SESSION_LIFETIME", 300))
+APP_ID = os.getenv("FLASK_EXAMPLE_APP_ID", "flask_example")
```
Comment (rngadam) [inline app.py]: flask-example to match the repo?

**Review by rngadam — CHANGES_REQUESTED** [review]
Pretty clean code, I like it.

Worth having a conversation about some of my questions here.

File: app.py (hunk 15 above)
Comment (k-allagbe) [reply app.py]: Done in config.py

File: app.py (hunk 13 above)
Comment (k-allagbe) [reply app.py]: Done (check `config.py`, `app_creator.py` and `main.py`). I will modify the package in membrane-backend to match. `configure_membrane` will have an `active` parameter that the user can set based on `MEMBRANE_SERVER`. `membrane_login_required` will check for auth only if `active` is set to `True`.

File: tests/test_routes.py (hunk 14 above)
Comment (k-allagbe) [reply tests/test_routes.py]: Fixed. It was meant to test that when membrane is not set, the route functions without redirection.

File: tests/test_routes.py (hunk 12 above)
Comment (k-allagbe) [reply tests/test_routes.py]: Done.

File: app.py (hunk 6 above)
Comment (k-allagbe) [reply app.py]: Done

File: app.py (hunk 7 above)
Comment (k-allagbe) [reply app.py]: Done. I extended flask-login. We might want to look into middlewares though, if we don't want to rely too much on other libraries.

File: config.py [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+DEFAULT_LOGGING_FORMAT = "%(asctime)s - %(name)s - %(levelname)s - %(message)s"
+DEFAULT_SESSION_LIFETIME = 300
+DEFAULT_TOKEN_EXPIRES_IN_SECONDS = 300
+DEFAULT_APP_ID = "flask-example"
+DEFAULT_REDIRECT_PATH = "/"
+
+load_dotenv()
+
+
+def read_key(membrane_server, env):
```
Comment (rngadam) [inline config.py]: read_membrane_key(membrane_server, env_variable)

File: config.py [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+    if membrane_server:
+        return open(os.getenv(env), "rb").read()
```
Comment (rngadam) [inline config.py]: https://docs.quantifiedcode.com/python-anti-patterns/maintainability/not_using_with_to_open_files.html

can you investigate if one of the linter could have caught this?

https://code.visualstudio.com/docs/python/linting

File: main.py [hunk 18]
```diff
@@ -0,0 +1,22 @@
+from flask import Blueprint, current_app, request
+from membrane.client.flask import membrane_current_user, membrane_login_required
+
+main = Blueprint("main", __name__)
```
Comment (rngadam) [inline main.py]: __name__ evaluates to the module. there's no module at the root level right? shouldn't this be placed in a module testapp/__init__.py to haev a reasonable name? 

main.py is a terrible name.

File: tests/test_routes.py [hunk 19]
```diff
@@ -1,41 +1,103 @@
...
+            self.assertEqual(response.status_code, 200)
+
+    def test_example_endpoint_with_login_redirect(self):
+        """Test that the / route redirects when a user is not logged in."""
+        with self.app.test_request_context():
+            response = self.client.get("/")
+            self.assertEqual(response.status_code, 302)
+            self.assertTrue(
+                response.headers["Location"].startswith(self.config.MEMBRANE_SERVER)
+            )
```
Comment (rngadam) [inline tests/test_routes.py:87]: no logout test?

**Review by rngadam — CHANGES_REQUESTED** [review]

File: config.py (hunk 16 above)
Comment (k-allagbe) [reply config.py]: Done

File: config.py (hunk 17 above)
Comment (k-allagbe) [reply config.py]: Not out of the box. They would need advanced configuration.

File: main.py (hunk 18 above)
Comment (k-allagbe) [reply main.py]: Done. Moved to routes/init.py.

File: tests/test_routes.py (hunk 19 above)
Comment (k-allagbe) [reply tests/test_routes.py:87]: Done. Added login and logout tests.

File: app.py [hunk 20]
```diff
@@ -1,4 +1,4 @@
-from app_creator import create_app
+from factory import create_app
```
Comment (rngadam) [inline app.py]: factory seems less informative than app_creator? can you explain this choice? will you stick every creation in this file?

**Review by rngadam — CHANGES_REQUESTED** [review]

File: config.py (hunk 17 above)
Comment (rngadam) [reply config.py]: adding https://docs.quantifiedcode.com/python-anti-patterns/index.html to developer training.

maybe something to investigate later

File: routes/__init__.py [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+
+@blueprint.before_request
+def log_request_info():
+    current_app.logger.debug("Headers: %s", request.headers)
+    current_app.logger.debug("Body: %s", request.get_data())
+
+
+@blueprint.route("/health", methods=["GET"])
+def health():
+    return "ok", 200
```
Comment (rngadam) [inline routes/__init__.py:15]: @RussellJimmies your nachet-backend /health should line up with this one. In general we want constants to be in variables configured in .env, but in this particular /health will evolve in the future to provide dynamic app health information (counters of exceptions HTTP codes 4xx, 3xx and 5xxx counts, etc)

File: app.py (hunk 20 above)
Comment (k-allagbe) [reply app.py]: As of now, nothing else is created with that file. `app_creator` is better.

**Review by rngadam — COMMENTED** [review]
Between the tests and the relentless changes, I think this is really nice. 

I'd probably want to test 

1) that the Membrane functionality actually works when deployed
2) test with a newbie developer if they are able to use it properly

@k-allagbe you asked for review but I see this is still draft. Can you make a checklist in the description of the PR to list what other changes (if any) you are planning to make? If not, maybe turn off the Draft flag?

Also would expect a README.md to complement the code.

**Comment by rngadam** [comment]
> Between the tests and the relentless changes, I think this is really nice.
> 
> I'd probably want to test
> 
> 1. that the Membrane functionality actually works when deployed
> 2. test with a newbie developer if they are able to use it properly
> 
> @k-allagbe you asked for review but I see this is still draft. Can you make a checklist in the description of the PR to list what other changes (if any) you are planning to make? If not, maybe turn off the Draft flag?
> 
> Also would expect a README.md to complement the code.

I think we should get this merged as this is supposed to be our application example.

**Review by vivalareda — APPROVED** [review]
looks good to me!

**Review by rngadam — APPROVED** [review]

**Review by RussellJimmies — APPROVED** [review]
Very clean, LGTM!

**Review by ThomasCardin — APPROVED** [review]
lgtm!

**Pull request merged**